src/main/java/org/junitpioneer/jupiter/AbstractEntryBasedExtension.java

```diff
@@ -0,0 +1,170 @@
+
+	private static final Namespace NAMESPACE = Namespace.create(AbstractEntryBasedExtension.class);
+	private static final String BACKUP = "Backup";
+
+	/**
+	 * @param context The current extension context.
+	 * @return <code>true</code> if one or more of the extension's annotations are present.
+	 */
+	protected abstract boolean isAnnotationPresent(ExtensionContext context);
+
+	/**
+	 * @param context The current extension context.
+	 * @return The entry keys to be cleared.
+	 */
+	protected abstract Set<K> entriesToClear(ExtensionContext context);
+
+	/**
+	 * @param context The current extension context.
+	 * @return The entry keys and values to be set.
+	 */
+	protected abstract Map<K, V> entriesToSet(ExtensionContext context);
+
+	/**
+	 * @return How to clear an entry based on its key.
+	 */
+	protected abstract Consumer<? super K> clearer();
+
+	/**
+	 * @return How to get an entry value based on its key.
+	 */
+	protected abstract Function<? super K, ? extends V> getter();
+
+	/**
+	 * @return How to set an entry value based on its key.
+	 */
+	protected abstract BiConsumer<? super K, ? super V> setter();
+
+	@Override
+	public void beforeAll(ExtensionContext context) throws Exception {
+		clearAndSetEntries(context);
+	}
+
+	@Override
+	public void beforeEach(ExtensionContext context) throws Exception {
+		clearAndSetEntries(context);
+	}
+
+	private void clearAndSetEntries(ExtensionContext context) {
+		Set<K> entriesToClear;
+		Map<K, V> entriesToSet;
+
+		try {
+			entriesToClear = entriesToClear(context);
+			entriesToSet = entriesToSet(context);
+			preventClearAndSetSameEntries(entriesToClear, entriesToSet.keySet());
+		}
+		catch (IllegalStateException ex) {
+			throw new ExtensionConfigurationException("Don't clear/set the same entry more than once.", ex);
+		}
+
+		if (entriesToClear.isEmpty() && entriesToSet.isEmpty())
+			return;
+
+		storeOriginalEntries(context, entriesToClear, entriesToSet.keySet());
+		clearEntries(entriesToClear);
+		setEntries(entriesToSet);
+	}
+
+	private void preventClearAndSetSameEntries(Collection<K> entriesToClear, Collection<K> entriesToSet) {
+		entriesToClear
+				.stream()
+				.filter(entriesToSet::contains)
+				.map(Object::toString)
+				.reduce((e0, e1) -> e0 + ", " + e1)
+				.ifPresent(duplicateEntries -> {
+					throw new IllegalStateException(
+						"Cannot clear and set the following entries at the same time: " + duplicateEntries);
+				});
+	}
+
+	private void storeOriginalEntries(ExtensionContext context, Collection<K> entriesToClear,
+			Collection<K> entriesToSet) {
+		context.getStore(NAMESPACE).put(BACKUP, new EntriesBackup(entriesToClear, entriesToSet));
+	}
+
+	private void clearEntries(Collection<K> entriesToClear) {
+		entriesToClear.forEach(clearer());
+	}
+
+	private void setEntries(Map<K, V> entriesToSet) {
+		entriesToSet.forEach(setter());
+	}
+
+	@Override
+	public void afterEach(ExtensionContext context) throws Exception {
+		if (isAnnotationPresent(context))
```

Comment:
I wonder if this check is necessary? If not, we could also get rid of the abstract `isAnnotationPresent` method.